typing/typecore.ml

```diff
@@ -2250,7 +2250,7 @@ let name_cases default lst =
 
 let unify_exp env exp expected_ty =
   let loc = proper_exp_loc exp in
-  unify_exp_types loc env exp.exp_type expected_ty
+  unify_exp_types ~exp:exp.exp_desc loc env exp.exp_type expected_ty
```

Comment:
I would prefer to have a `try .. with` around the call that catches the specific exception we want to enrich with the `exp` metadata, rather than changing the prototype of `unify_exp_types`. Again, the idea would be to change only the places that are known to have a clear expression to designate for the error -- `unify_exp` is such a place, but the more general `unify_exp_types` isn't.

Reply:
Did that.